rpc/export.go

```diff
@@ -154,26 +155,75 @@ func (c *lockedConn) sendCap(d rpccp.CapDescriptor, client capnp.Client) (_ expo
 	state.Metadata.Lock()
 	defer state.Metadata.Unlock()
 	id, ok := c.findExportID(state.Metadata)
+	var ee *expent
 	if ok {
-		ent := c.lk.exports[id]
-		ent.wireRefs++
-		d.SetSenderHosted(uint32(id))
-		return id, true, nil
+		ee = c.lk.exports[id]
+		ee.wireRefs++
+	} else {
+		// Not already present; allocate an export id for it:
+		ee = &expent{
+			client:    client.AddRef(),
+			wireRefs:  1,
+			isPromise: state.IsPromise,
+		}
+		id = exportID(c.lk.exportID.next())
+		if int64(id) == int64(len(c.lk.exports)) {
+			c.lk.exports = append(c.lk.exports, ee)
+		} else {
+			c.lk.exports[id] = ee
+		}
+		c.setExportID(state.Metadata, id)
 	}
+	if ee.isPromise {
+		// Send a promise, wait for the resolution asynchronously, then send
+		// a resolve message:
+		d.SetSenderPromise(uint32(id))
+		waitRef := client.AddRef()
+		go func() {
+			// Logically we don't hold the lock anymore; it's held by the
+			// goroutine that spawned this one. So cast back to an unlocked
+			// Conn before trying to use it again:
+			unlockedConn := (*Conn)(c)
+
+			waitErr := waitRef.Resolve(c.bgctx)
```

Comment:
Note: it might be better to use context.WithCancel and rig up the CancelFunc to be called if the entry is dropped. Otherwise there's a risk of leaking goroutines in a workflow where promises are created and never resolved.

Reply:
Fixed this.